### PR TITLE
Version 4.0.0

### PR DESCRIPTION
**Pull request by walter-weinmann**

### CODE REVIEW

File: src/sqlparse.yrl [hunk 1]
```diff
@@ -831,7 +831,7 @@ target -> parameter_ref : '$1'.
 %% query expressions
 %% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
-query_exp ->                     query_term : '$1'.
+query_exp ->                         query_term     : '$1'.
```
Comment (c-bik) [inline src/sqlparse.yrl]: Fix indent align `:`

File: include/sqlparse_params_filter.hrl [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -----------------------------------------------------------------------------
+
+-ifndef(SQLPARSE_PARAMS_FILTER_HRL).
+-define(SQLPARSE_PARAMS_FILTER_HRL, true).
+
+-define(LAYOUT_INIT(_FunState, Ctx, PTree, _FoldState),
```
[the remark this answers is not in the document]
Comment (walter-weinmann) [reply include/sqlparse_params_filter.hrl]: Done.

File: src/sqlparse_identifiers_match.erl [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+% {atom, binary, _}
+% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+fold(InFields, _FunState, Ctx, {Elem1, Elem2, _Elem3} = _PTree, FoldState)
+    when is_atom(Elem1), is_binary(Elem2), element(2, FoldState) == start ->
+    ?LAYOUT_INIT(_FunState, Ctx, _PTree, FoldState),
+    RT = case lists:member(Elem2, InFields) of
+             true -> [Elem2 | Ctx];
+             _ -> Ctx
+         end,
```
[the remark this answers is not in the document]
Comment (walter-weinmann) [reply src/sqlparse_identifiers_match.erl]: Done.

File: include/sqlparse_identifiers_match.hrl [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+-ifndef(SQLPARSE_IDENTIFIERS_MATCH_HRL).
+-define(SQLPARSE_IDENTIFIERS_MATCH_HRL, true).
+
+-define(LAYOUT_INIT(_FunState, Ctx, PTree, _FoldState),
+    ?debugFmt(
+        "~p:~p ===> Start~n FunState: ~p~n CtxIn: ~p~n PTree: ~p~n FoldState: ~p~n",
+        [?MODULE, ?FUNCTION_NAME, _FunState, Ctx, PTree, _FoldState])).
+-define(LAYOUT_RESULT(RT),
+    ?debugFmt("~p:~p ===>~n CtxOut: ~p~n", [?MODULE, ?FUNCTION_NAME, RT]),
+    RT).
```
[the remark this answers is not in the document]
Comment (walter-weinmann) [reply include/sqlparse_identifiers_match.hrl]: Done.

File: src/sqlparse_identifiers_match.erl [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+-include("sqlparse_identifiers_match.hrl").
+
+% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Setting up parameters.
+% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+-spec init(InFields :: [binary()]) -> InFields :: [binary()].
+init(InFields)
+    when is_list(InFields) ->
+    ?debugFmt("~p:~p ===> Start~n InFields: ~p~n",
```
[the remark this answers is not in the document]
Comment (walter-weinmann) [reply src/sqlparse_identifiers_match.erl]: Done.

File: src/sqlparse_identifiers_match.erl [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+         end,
+    ?LAYOUT_RESULT(RT);
+
+% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% {atom, binary}
+% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+fold(InFields, _FunState, Ctx, {Elem1, Elem2} = _PTree, FoldState)
+    when is_atom(Elem1), Elem1 =/= extra, is_binary(Elem2),
+    element(2, FoldState) == start ->
```
[the remark this answers is not in the document]
Comment (walter-weinmann) [reply src/sqlparse_identifiers_match.erl]: Done - whole project.

File: include/sqlparse.hrl [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -----------------------------------------------------------------------------
+
+-ifndef(SQLPARSE_HRL).
+-define(SQLPARSE_HRL, true).
+
+-define(D(Format), ?D(Format, [])).
+-define(D(Format, Args), ?debugFmt("~p:~p ===> " ++ Format,
```
Comment (c-bik) [inline include/sqlparse.hrl]: `++` isn’t required

File: rebar.config.script [hunk 8]
```diff
@@ -1,7 +1,7 @@
...
-        lists:keystore(coveralls_service_job_id, 1, CONFIG, {coveralls_service_job_id, JobId});
-    _ ->
-        CONFIG
-end.
+"true" ->
+JobId = os:getenv("TRAVIS_JOB_ID"),
+lists:keystore(coveralls_service_job_id, 1, CONFIG, {coveralls_service_job_id, JobId});
+_ ->
+CONFIG
+end .
```
Comment (c-bik) [inline rebar.config.script]: Remove space between `end` and `.`

File: src/sqlparse_identifiers_match.erl [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+-export([
+    finalize/2,
+    fold/5,
+    init/1
+]).
+
+-define(NODEBUG, true).
+
+-include_lib("eunit/include/eunit.hrl").
```
Comment (c-bik) [inline src/sqlparse_identifiers_match.erl]: This makes this module looks like it only contains eunit tests. When actually there are no eunit tests in it. Why do we need this here?

Comment (walter-weinmann) [reply src/sqlparse_identifiers_match.erl]: ?debugFmt.

File: include/sqlparse.hrl (hunk 7 above)
Comment (walter-weinmann) [reply include/sqlparse.hrl]: Done.

File: src/sqlparse.yrl [hunk 10]
```diff
@@ -1255,19 +1255,14 @@ Erlang code.
...
-    pt_to_string_bu/1,
-    pt_to_string_format/1
+    parsetree_with_tokens/1
 ]).
 
 -define(NODEBUG, true).
--include_lib("eunit/include/eunit.hrl").
+
+-include("sqlparse.hrl").
 -include("sqlparse_fold.hrl").
```
Comment (c-bik) [inline src/sqlparse.yrl]: Do we still need to include `sqlparse_fold.hrl` here?

File: include/sqlparse_fold.hrl [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    "nvarchar2",
+    "raw",
+    "rowid",
+    "timestamp",
+    "urowid",
+    "varchar2"
+]).
+
+-define(FOLD_INIT(FunState, Ctx, PTree),
+    ?D("Start~n FunState: ~p~n CtxIn: ~p~n PTree: ~p~n",
```
Comment (c-bik) [inline include/sqlparse_fold.hrl]: Include `sqlparse.hrl` here!

**Pull request merged**